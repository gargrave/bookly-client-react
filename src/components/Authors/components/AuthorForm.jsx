import React from 'react'
import { func, shape, string } from 'prop-types'

const AuthorForm = props => {
  return (
    <form noValidate style={{ maxWidth: '500px', margin: 'auto', marginTop: '20px', textAlign: 'left' }}>
      <div className="input-single">
        <label htmlFor="firstName">First name:</label>
        <input
          type="text"
          name="firstName"
          id="firstName"
          value={props.author.firstName}
          onChange={props.handleInputChange}
        />
      </div>

      <div className="input-single">
        <label htmlFor="lastName">Last name:</label>
        <input
          type="text"
          name="lastName"
          id="lastName"
          value={props.author.lastName}
          onChange={props.handleInputChange}
        />
      </div>

      <div className="input-single">
        <button type="button" onClick={props.handleSubmit}>
          Submit
        </button>
        <button type="button" className="button-info float-right" onClick={props.handleCancel}>
          Cancel
        </button>
      </div>
    </form>
  )
}

AuthorForm.propTypes = {
  author: shape({
    firstName: string.isRequired,
    lastName: string.isRequired
  }).isRequired,
  handleInputChange: func.isRequired,
  handleSubmit: func.isRequired,
  handleCancel: func.isRequired
}

export default AuthorForm
