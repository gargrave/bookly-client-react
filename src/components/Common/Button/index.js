import React from 'react'
import { func, string } from 'prop-types'

const classname = (type) => {
  if (type === 'success') {
    return 'buttons-success'
  }
}

const Button = ({
  onClick,
  text,
  type,
}) => {
  return (
    <button
      className={classname(type)}
      style={{ marginLeft: '10px' }}
      onClick={onClick}>
      { text }
    </button>
  )
}

Button.propTypes = {
  onClick: func.isRequired,
  text: string.isRequired,
  type: string.isRequired,
}

export default Button
