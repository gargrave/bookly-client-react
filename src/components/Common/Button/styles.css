.bookly-button {
  margin-left: 10px;
}
