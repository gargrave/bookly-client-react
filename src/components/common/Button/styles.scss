.bookly-button {
  &:not(&--left) {
    margin-left: 10px;
  }
}
