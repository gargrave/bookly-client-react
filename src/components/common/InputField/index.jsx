// @flow
import React from 'react';
import { func, oneOf, string } from 'prop-types';

type Props = {
  boundValue: string,
  label: string,
  name: string,
  onInputChange: Function,
  type?: string,
};

const acceptableTypes = [
  'email',
  'password',
  'text',
];

function InputField({
  boundValue,
  label,
  name,
  onInputChange,
  type,
}: Props) {
  return (
    <div className="input-field">
      <label htmlFor={name}>{label}:</label>
      <input
        type={type || 'text'}
        name={name}
        id={name}
        value={boundValue}
        onChange={onInputChange} />
    </div>
  );
}

InputField.propTypes = {
  boundValue: string.isRequired,
  label: string.isRequired,
  name: string.isRequired,
  onInputChange: func.isRequired,
  type: oneOf(acceptableTypes),
};

export default InputField;
