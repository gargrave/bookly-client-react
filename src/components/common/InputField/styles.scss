.bookly-input-field {
  &__input {
    &--invalid {
      border-color: #ff4136 !important; // using !important to override Shoelace's styles
    }
  }

  &__error {
    color: #ff4136;
  }
}
