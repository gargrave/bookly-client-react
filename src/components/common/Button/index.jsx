// @flow
import React from 'react';
import { bool, func, oneOf, string } from 'prop-types';

import { buildClasses } from '../../../utils/cssHelpers';

import './styles.css';

type Props = {
  canSubmit?: boolean,
  classes?: string,
  onClick: Function,
  position?: string,
  text: string,
  type: string,
};

const acceptableTypes = [
  'success',
  'secondary',
  'info',
  'warning',
  'danger',
  'light',
  'dark',
];

const acceptablePositions = [
  'left',
];

function buttonClass(type: string) {
  if (acceptableTypes.includes(type)) {
    return `button-${type}`;
  }
  return '';
}

function positionClass(position: string) {
  if (acceptablePositions.includes(position)) {
    return `button-${position}`;
  }
  return '';
}

function buildClassList(
  type: string,
  position: string,
  classes: string,
) {
  return buildClasses(
    ['button', positionClass(position)],
    [buttonClass(type), ...classes.split(' ')]
  );
}

function Button({
  canSubmit,
  classes = '',
  onClick,
  position = '',
  text,
  type,
}: Props) {
  return (
    <button
      type={canSubmit ? 'submit' : 'button'}
      className={buildClassList(type, position, classes)}
      onClick={onClick}>
      { text }
    </button>
  );
}

Button.propTypes = {
  canSubmit: bool,
  classes: string,
  onClick: func.isRequired,
  position: oneOf(acceptablePositions),
  text: string.isRequired,
  type: oneOf(acceptableTypes).isRequired,
};

export default Button;
