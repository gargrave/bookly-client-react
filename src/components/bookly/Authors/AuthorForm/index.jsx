import React from 'react'
import { func, shape, string } from 'prop-types'

import { buildClasses } from '@/utils/cssHelpers'

import Button from '@/components/common/Button'
import InputField from '@/components/common/InputField'

import '@/components/common/Form/styles.css'

const AuthorForm = ({
  author,
  onCancel,
  onInputChange,
  onSubmit,
}) => {
  return (
    <form
      className={buildClasses(['form', 'author-form'])}
      onSubmit={onSubmit}
      noValidate>

      <InputField
        boundValue={author.firstName}
        label="First name"
        name="firstName"
        onInputChange={onInputChange} />

      <InputField
        boundValue={author.lastName}
        label="Last name"
        name="lastName"
        onInputChange={onInputChange} />

      <div className="input-field">
        <Button
          canSubmit={true}
          onClick={onSubmit}
          position="left"
          text="Submit"
          type="success" />
        <Button
          extraClasses="float-right"
          onClick={onCancel}
          text="Cancel"
          type="info" />
      </div>
    </form>
  )
}

AuthorForm.propTypes = {
  author: shape({
    firstName: string.isRequired,
    lastName: string.isRequired,
  }).isRequired,
  onCancel: func.isRequired,
  onInputChange: func.isRequired,
  onSubmit: func.isRequired,
}

export default AuthorForm
