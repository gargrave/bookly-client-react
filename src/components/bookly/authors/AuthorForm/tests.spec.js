import React from 'react'
import { shallow } from 'enzyme'

import { authorsMockData } from '@/utils/mocks/data/authors-mock-data'

import Button from '@/components/common/Button'
import InputField from '@/components/common/InputField'

import AuthorForm from './'

describe('AuthorForm', () => {
  let props
  let component

  describe('with "author" populated', () => {
    beforeEach(() => {
      props = {
        author: Object.create(authorsMockData[0]),
        onInputChange: jest.fn(),
        onSubmit: jest.fn(),
        onCancel: jest.fn(),
      }

      component = shallow(<AuthorForm {...props} />)
    })

    test('renders correctly', () => {
      expect(component).toMatchSnapshot()
      expect(component.find('.bookly-form.bookly-author-form').length).toEqual(1)
      expect(component.find(Button).length).toEqual(2)
      expect(component.find(InputField).length).toEqual(2)
    })
  })
})
