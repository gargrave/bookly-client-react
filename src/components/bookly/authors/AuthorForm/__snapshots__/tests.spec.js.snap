// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorForm with "author" populated renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthorForm
    author={Object {}}
    onCancel={[Function]}
    onInputChange={[Function]}
    onSubmit={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <InputField
          boundValue="Aldous"
          label="First name"
          name="firstName"
          onInputChange={[Function]}
        />,
        <InputField
          boundValue="Huxley"
          label="Last name"
          name="lastName"
          onInputChange={[Function]}
        />,
        <div
          className="input-field"
        >
          <Button
            canSubmit={true}
            onClick={[Function]}
            position="left"
            text="Submit"
            type="success"
          />
          <Button
            extraClasses="float-right"
            onClick={[Function]}
            text="Cancel"
            type="info"
          />
        </div>,
      ],
      "className": "bookly-form bookly-author-form",
      "noValidate": true,
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "boundValue": "Aldous",
          "label": "First name",
          "name": "firstName",
          "onInputChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "boundValue": "Huxley",
          "label": "Last name",
          "name": "lastName",
          "onInputChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Button
              canSubmit={true}
              onClick={[Function]}
              position="left"
              text="Submit"
              type="success"
            />,
            <Button
              extraClasses="float-right"
              onClick={[Function]}
              text="Cancel"
              type="info"
            />,
          ],
          "className": "input-field",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "canSubmit": true,
              "onClick": [Function],
              "position": "left",
              "text": "Submit",
              "type": "success",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "extraClasses": "float-right",
              "onClick": [Function],
              "text": "Cancel",
              "type": "info",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <InputField
            boundValue="Aldous"
            label="First name"
            name="firstName"
            onInputChange={[Function]}
          />,
          <InputField
            boundValue="Huxley"
            label="Last name"
            name="lastName"
            onInputChange={[Function]}
          />,
          <div
            className="input-field"
          >
            <Button
              canSubmit={true}
              onClick={[Function]}
              position="left"
              text="Submit"
              type="success"
            />
            <Button
              extraClasses="float-right"
              onClick={[Function]}
              text="Cancel"
              type="info"
            />
          </div>,
        ],
        "className": "bookly-form bookly-author-form",
        "noValidate": true,
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "boundValue": "Aldous",
            "label": "First name",
            "name": "firstName",
            "onInputChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "boundValue": "Huxley",
            "label": "Last name",
            "name": "lastName",
            "onInputChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Button
                canSubmit={true}
                onClick={[Function]}
                position="left"
                text="Submit"
                type="success"
              />,
              <Button
                extraClasses="float-right"
                onClick={[Function]}
                text="Cancel"
                type="info"
              />,
            ],
            "className": "input-field",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "canSubmit": true,
                "onClick": [Function],
                "position": "left",
                "text": "Submit",
                "type": "success",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "extraClasses": "float-right",
                "onClick": [Function],
                "text": "Cancel",
                "type": "info",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
